presto-main/src/main/java/io/prestosql/operator/PageBuffer.java

```diff
@@ -53,6 +53,10 @@ public PageBuffer(ListenableFuture<?> initialFuture)
     public WorkProcessor<Page> pages()
     {
         return WorkProcessor.create(() -> {
+            if (isFinished() && isEmpty()) {
```

Comment:
What does `isEmpty` mean here?  I'm concerned that this may short circuit things that don't have output but create sideeffects for others (e.g., hash build).

Reply:
`isFinished()` essentially means EOS, so no more input pages are expected. However, `PageBuffer` is not really finished until buffered page is consumed (that's why we need `isEmpty()` check).
Here, we favor early finish over initial blocking (that's the old join behavior).

I added tests that validate expected behavior.

Note: instead of blocking on `lookupSourceProviderFuture` we could initially block on `dynamicFilterFuture`. Thus we wouldn't wait for entire build side before opening table scan split, but just for enough data to fill up dynamic filter. However, build side could be result of aggregation which requires entire table to be read. It's a tricky question: do we favor probe side short circuit over holding split while waiting for dynamic filter?

Reply:
I'm not sure what we want to do in the long run, but this seems good for now.